specs/monitor_ceph_pool_utilization.adoc

```diff
@@ -0,0 +1,190 @@
+= Collect and process ceph pool utlization trending data
+
+The performance monitoring module if enabled in Tendrl, should collect and
+process the ceph pool utilization data in time series DB. This would help in
+creating trending charts in Tendrl UI which can help storage administrator to
+decide on requirement for additional storage in the system.
+
+== Problem description
+
+If ceph pool utilization trending is not available to administrator, they cannot
+easily decide how the pools are getting filled up and at what point in time more
+additional storage to be brought in and provisioned in the system.
+
+== Use Cases
+
+This PR addresses the trending charts use case in Tendrl for the ceph storage
+pools.
+
+== Proposed change
+
+* Introduce a collectd plugin which collects the ceph pool utilization data from
+underlying cluster and pushes it to time series DB at regular intervals.
+
+* Enhance the monitoring module to set required thresholds for the ceph pool
+utilization
+
+* Enhance the monitoring module to frame and raise required alerts while breach
+of threshold limits
+
+=== Alternatives
+
+* There are different ways to achieve the reading of pool utilization data
+
+* Option-1: Use rados module directly to invoke mon_command `df` to fetch the
+pool utilization data. A connection to rados could be kept open and keep running
+`df` command for getting details
+
+```
+import time
+import rados
+import json
+
+def get_pool_utilization(r):
+    cmd = {"prefix":"df", "format":"json"}
+    ret, buf, errs = r.mon_command(json.dumps(cmd), b'', timeout=5)
+    result = json.loads(buf)
+    return result
+
+try:
+    r = rados.Rados(conffile="/etc/ceph/ceph.conf")
+    r.connect()
+
+    for i in range(1000):
+        print(get_pool_utilization(r))
+        time.sleep(1)
+except KeyboardInterrupt:
+    pass
+finally:
+    if r is not None:
+        r.shutdown()
+```
+
+* Option-2: The other option is to use execute command `ceph df` as a process
+and parse the output for data
+
+```
+import json
+import subprocess
+
+cluster = 'my_cluster'
+cmd = [
+    'ceph',
+    '--cluster',
+    cluster,
+    '-f',
+    'json-pretty',
+    'df'
+]
+process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+out, err = process.communicate()
+ret_code = process.poll()
+if ret_code:
+    print "command %s exited with non-zero status: %d" % (cmd, ret_code)
+    return
+
+df = json.loads(out)
+```
+
+=== Data model impact:
+
+None
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+* API module to support the fetching of ceph pool utilization data as well as
+part of listing generic monitoring data using resources
+
```

Comment:
Probably, this needs to be either of the below:

1. The pool object in tendrl ceph integration definitions, will provide a way to indicate a way to read the utilization data from some directory in etcd if monitoring is enabled and the tendrl-api will look into this path to provide the utilization data
or
2. The performance-monitoring application will periodically update the utilization data to a location in etcd and also the performance monitoring application will expose apis to access this data and the tendrl-api will proxy to it.

I personally think that approach 2 although is a 2-step process, is safer and better than approach 1 of involving changes to definitions on the fly once the monitoring is enabled.